src/ray/raylet/agent_manager.h

```diff
@@ -32,96 +30,50 @@ typedef std::function<std::shared_ptr<boost::asio::deadline_timer>(std::function
                                                                    uint32_t delay_ms)>
     DelayExecutorFn;
 
-typedef std::function<std::shared_ptr<rpc::RuntimeEnvAgentClientInterface>(
-    const std::string &ip_address, int port)>
-    RuntimeEnvAgentClientFactoryFn;
-
-/// Callback that's called after runtime env is created.
-/// \param[in] successful Whether or not the creation was successful.
-/// \param[in] serialized_runtime_env_context Serialized context.
-/// \param[in] setup_error_message The error message if runtime env creation fails.
-/// It must be only set when successful == false.
-typedef std::function<void(bool successful,
-                           const std::string &serialized_runtime_env_context,
-                           const std::string &setup_error_message)>
-    GetOrCreateRuntimeEnvCallback;
-typedef std::function<void(bool successful)> DeleteRuntimeEnvIfPossibleCallback;
-
-class AgentManager : public rpc::AgentManagerServiceHandler {
+// Manages a separate "Agent" process. In constructor (or the `StartAgnet` method) it
+// starts a process with `agent_commands` plus some additional arguments.
+//
+// Raylet outlives agents. This means when Raylet exits, the agents must exit; on the
+// other hand, if the agent exits, the raylet may exit only if fate_shares = true. This is
+// implement by the dtor of AgentManager to kill the agent process.
+//
+// We typically start these agents:
+// - The DashboardAgent: `ray/dashboard/agent.py`
+// - The RuntimeEnvAgent: `ray/_private/runtime_env/agent/main.py`
+class AgentManager {
  public:
   struct Options {
     const NodeID node_id;
+    const std::string agent_name;
+    // Commands to start the agent. Note we append extra arguments:
+    // --agent-id $AGENT_ID # A random string of int
     std::vector<std::string> agent_commands;
+    // If true: the started process fate-shares with the raylet. i.e. when the process
+    // fails to start or exits, we SIGTERM the raylet.
+    bool fate_shares;
   };
 
   explicit AgentManager(Options options,
                         DelayExecutorFn delay_executor,
-                        RuntimeEnvAgentClientFactoryFn runtime_env_agent_client_factory,
                         bool start_agent = true /* for test */)
       : options_(std::move(options)),
         delay_executor_(std::move(delay_executor)),
-        runtime_env_agent_client_factory_(std::move(runtime_env_agent_client_factory)) {
+        kill_raylet_if_process_exits_(options_.fate_shares) {
```

Comment:
nit, but we can just call it fate_shares?

Reply:
The awkward point is, in dtor I changed this bool to false to avoid killing raylet in the shutdown process. Do you think this can be considered that "raylet no longer fate-shares with the process after the dtor killing" ? If so I can rename it.

Reply:
Actually I feel like fate_share kinda sounds clear in this case (we stop fate sharing when we stop the monitor thread). Can you also add an inline comment "stop fate sharing because we gracefully kill the agent".

Reply:
done.